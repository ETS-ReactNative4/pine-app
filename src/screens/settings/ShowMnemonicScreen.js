import React, { Component } from 'react';
import { StyleSheet, Alert } from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import * as keyActions from '../../actions/keys';
import getMnemonicByKey from '../../crypto/getMnemonicByKey';
import headerStyles from '../../styles/headerStyles';
import MnemonicWordsContainer from '../../containers/MnemonicWordsContainer';
import BackButton from '../../components/BackButton';
import SettingsAttribute from '../../components/SettingsAttribute';
import SettingsGroup from '../../components/SettingsGroup';
import SettingsDescription from '../../components/SettingsDescription';
import BaseSettingsScreen from './BaseSettingsScreen';

const styles = StyleSheet.create({
  view: {
    padding: 20,
    overflow: 'hidden'
  },
  blur: {
    top: -20,
    bottom: -20,
    left: -10,
    right: -10
  }
});

@connect((state) => ({
  keys: state.keys.items
}))
export default class ShowMnemonicScreen extends Component {
  static navigationOptions = ({ navigation }) => ({
    title: 'Recovery Key',
    headerStyle: headerStyles.header,
    headerTitleStyle: headerStyles.title,
    headerLeft: (<BackButton onPress={() => { navigation.goBack(); }} />)
  });

  state = {
    phrase: '',
    storeInICloud: false
  }

  componentDidMount() {
    const dispatch = this.props.dispatch;
    const keys = Object.values(this.props.keys);
    const defaultKey = keys[0];

    getMnemonicByKey(defaultKey.id).then((mnemonic) => {
      this.setState({ phrase: mnemonic });

      return dispatch(keyActions.recover()).then((recoveredMnemonic) => {
        if (recoveredMnemonic !== mnemonic) {
          return;
        }

        this.setState({ storeInICloud: true });
      });
    });
  }

  _updateICloudState(storeInICloud) {
    const dispatch = this.props.dispatch;
    const mnemonic = this.state.phrase;

    this.setState({ storeInICloud });

    if (storeInICloud) {
      return dispatch(keyActions.backup(mnemonic));
    }

    return dispatch(keyActions.removeBackup());
  }

  _onStoreInICloudChange(storeInICloud) {
    const activate = !this.state.storeInICloud && storeInICloud;
    let title;
    let message;
    let buttonText;
    let buttonStyle;

    if (activate) {
      title = 'Back up in iCloud?';
      message = 'Saving your recovery key in your iCloud account is potentially less secure than writing it down and storing it yourself.';
      buttonText = 'Back up in iCloud';
      buttonStyle = 'default';
    } else {
      title = 'Remove from iCloud?';
      message = 'Only remove your recovery key from iCloud if you have written it down and stored it yourself first.';
      buttonText = 'Remove from iCloud';
      buttonStyle = 'destructive';
    }

    Alert.alert(
      title,
      message,
      [
        { text: 'Cancel', style: 'cancel' },
        {
          text: buttonText,
          style: buttonStyle,
          onPress: this._updateICloudState.bind(this, storeInICloud)
        }
      ],
      { cancelable: false }
    );
  }

  render() {
    return (
      <BaseSettingsScreen>
        <SettingsGroup style={styles.view}>
          <MnemonicWordsContainer phrase={this.state.phrase} blurStyle={styles.blur} />
        </SettingsGroup>
        <SettingsDescription>
          Write down and store this recovery key in a safe place so you can recover
          your wallet if you would lose or break your phone.
        </SettingsDescription>

        <SettingsGroup>
          <SettingsAttribute
            name='Back up in iCloud'
            value={this.state.storeInICloud}
            onValueChange={this._onStoreInICloudChange.bind(this)}
            isLastItem={true}
          />
        </SettingsGroup>
        <SettingsDescription>
          Saving your recovery key in your iCloud account is potentially less secure than writing it down and storing it yourself.
        </SettingsDescription>
      </BaseSettingsScreen>
    );
  }
}

ShowMnemonicScreen.propTypes = {
  dispatch: PropTypes.func,
  navigation: PropTypes.any,
  keys: PropTypes.object
};
