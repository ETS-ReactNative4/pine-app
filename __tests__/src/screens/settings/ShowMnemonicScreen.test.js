import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import ShowMnemonicScreen from '../../../../src/screens/settings/ShowMnemonicScreen';

jest.mock('../../../../src/containers/ErrorModalContainer', () => 'ErrorModalContainer');

const storeMock = {
  getState: jest.fn(() => ({
    keys: {
      items: {
        'b62a06e9-7a1f-4862-a350-7af4505114c7': {
          id: 'b62a06e9-7a1f-4862-a350-7af4505114c7'
        }
      }
    }
  })),
  dispatch: jest.fn(),
  subscribe: jest.fn()
};

const navigationMock = {
  navigate: jest.fn(),
  state: {
    params: {}
  }
};

describe('ShowMnemonicScreen', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <ShowMnemonicScreen store={storeMock} navigation={navigationMock} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
