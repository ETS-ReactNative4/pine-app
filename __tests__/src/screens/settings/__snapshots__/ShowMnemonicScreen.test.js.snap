// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowMnemonicScreen renders correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#EFEFF3",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "paddingTop": 35,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "borderBottomColor": "#C8C8CC",
                "borderBottomWidth": 0.5,
                "borderTopColor": "#C8C8CC",
                "borderTopWidth": 0.5,
                "marginBottom": 35,
              },
              Object {
                "overflow": "hidden",
                "padding": 20,
              },
            ]
          }
        >
          <MnemonicWordsContainer
            blurStyle={
              Object {
                "bottom": -20,
                "left": -10,
                "right": -10,
                "top": -20,
              }
            }
            phrase=""
          />
        </View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={undefined}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 30,
                "marginTop": -25,
                "paddingLeft": 15,
                "paddingRight": 15,
              },
            ]
          }
        >
          Write down and store this recovery key in a safe place so you can recover your wallet if you would lose or break your phone.
        </Text>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "borderBottomColor": "#C8C8CC",
                "borderBottomWidth": 0.5,
                "borderTopColor": "#C8C8CC",
                "borderTopWidth": 0.5,
                "marginBottom": 35,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderBottomColor": "#C8C7CC",
                  "borderBottomWidth": 0.5,
                  "flex": 1,
                  "height": 44,
                  "justifyContent": "center",
                  "marginLeft": 15,
                  "paddingRight": 15,
                },
                Object {
                  "borderBottomWidth": 0,
                },
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              numberOfLines={undefined}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "fontSize": 17,
                    "fontWeight": "400",
                    "lineHeight": 22,
                  },
                ]
              }
            >
              Back up in iCloud
            </Text>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#8F8E94",
                    "fontSize": 17,
                    "fontWeight": "400",
                    "lineHeight": 22,
                    "position": "absolute",
                    "right": 15,
                    "textAlign": "right",
                    "width": 250,
                  },
                  Object {
                    "width": null,
                  },
                ]
              }
            >
              <RCTSwitch
                disabled={false}
                onChange={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                onValueChange={[Function]}
                style={
                  Array [
                    Object {
                      "height": 31,
                      "width": 51,
                    },
                    undefined,
                  ]
                }
                value={false}
              />
            </View>
          </View>
        </View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={undefined}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 30,
                "marginTop": -25,
                "paddingLeft": 15,
                "paddingRight": 15,
              },
            ]
          }
        >
          Saving your recovery key in your iCloud account is potentially less secure than writing it down and storing it yourself.
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <ErrorModalContainer />
</View>
`;
